src/outoforder.cpp

```diff
@@ -0,0 +1,144 @@
+// Copyright (c) 2018 Blockstream
+
+#include <chain.h>
+#include <chainparams.h>
+#include <consensus/validation.h>
+#include <dbwrapper.h>
+#include <primitives/block.h>
+#include <sync.h>
+#include <uint256.h>
+#include <util.h>
+#include <validation.h>
+
+#include <deque>
+#include <map>
+#include <memory>
+#include <utility>
+#include <vector>
+
+static int ExtractHeightFromBlock(const Consensus::Params& consensusParams, const std::shared_ptr<const CBlock> pblock)
+{
+    if (pblock->nBits > consensusParams.BIP34AssumedBits) {
+        // We can't be sure BIP34 is active, and without that we can't figure out the height yet
+        return -1;
+    }
+
+    if (pblock->vtx.size() < 1) return -1;
+    if (pblock->vtx[0]->vin.size() < 1) return -1;
+
+    const auto& scriptSig = pblock->vtx[0]->vin[0].scriptSig;
+    auto pc = scriptSig.begin();
+    opcodetype opcode;
+    std::vector<unsigned char> data;
+    if (!scriptSig.GetOp(pc, opcode, data)) return -1;
+    try {
+        CScriptNum bn(data, /*require minimal encoding=*/true);
+        return bn.getint();
+    }
+    catch (scriptnum_error) {
+        return -1;
+    }
+}
+
+static const char DB_SUBSEQUENT_BLOCK = 'S';
+
+static CCriticalSection cs_ooob;
+
+static CDBWrapper* GetOoOBlockDB() EXCLUSIVE_LOCKS_REQUIRED(cs_ooob)
+{
+    static CDBWrapper *ooob_db = nullptr;
+    if (!ooob_db) ooob_db = new CDBWrapper(GetDataDir() / "future_blocks", /*cache size=*/1024);
+    return ooob_db;
+}
+
+bool StoreOoOBlock(const CChainParams& chainparams, const std::shared_ptr<const CBlock> pblock) EXCLUSIVE_LOCKS_REQUIRED(cs_main)
```

Comment:
It'd be nice to put a CheckBlock in here first.

Reply:
Put it at the caller, since that's where it's deciding whether or not to do the storing.